mesonbuild/interpreter/interpreter.py

```diff
@@ -3128,27 +3129,37 @@ def validate_installable_file(fpath: Path) -> bool:
             # subproject files, as long as they are scheduled to be installed.
             if validate_installable_file(norm):
                 return
-        norm = Path(os.path.abspath(Path(srcdir, subdir, fname)))
-        if os.path.isdir(norm):
-            inputtype = 'directory'
-        else:
-            inputtype = 'file'
-        if InterpreterRuleRelaxation.ALLOW_BUILD_DIR_FILE_REFERENCES in self.relaxations and builddir in norm.parents:
-            return
-        if srcdir not in norm.parents:
-            # Grabbing files outside the source tree is ok.
-            # This is for vendor stuff like:
-            #
-            # /opt/vendorsdk/src/file_with_license_restrictions.c
-            return
-        project_root = Path(srcdir, self.root_subdir)
-        subproject_dir = project_root / self.subproject_dir
-        if norm == project_root:
+
+        def do_validate_within_subproject(norm):
+            if os.path.isdir(norm):
+                inputtype = 'directory'
+            else:
+                inputtype = 'file'
+            if InterpreterRuleRelaxation.ALLOW_BUILD_DIR_FILE_REFERENCES in self.relaxations and builddir in norm.parents:
+                return
+            if srcdir not in norm.parents:
+                # Grabbing files outside the source tree is ok.
+                # This is for vendor stuff like:
+                #
+                # /opt/vendorsdk/src/file_with_license_restrictions.c
+                return
+            project_root = Path(srcdir, self.root_subdir)
+            subproject_dir = project_root / self.subproject_dir
+            if norm == project_root:
+                return
+            if project_root not in norm.parents:
+                raise InterpreterException(f'Sandbox violation: Tried to grab {inputtype} {norm.name} outside current (sub)project.')
+            if subproject_dir == norm or subproject_dir in norm.parents:
+                raise InterpreterException(f'Sandbox violation: Tried to grab {inputtype} {norm.name} from a nested subproject.')
+
+        fname = os.path.join(subdir, fname)
+        if fname in self.validated_cache:
             return
-        if project_root not in norm.parents:
-            raise InterpreterException(f'Sandbox violation: Tried to grab {inputtype} {norm.name} outside current (sub)project.')
-        if subproject_dir == norm or subproject_dir in norm.parents:
-            raise InterpreterException(f'Sandbox violation: Tried to grab {inputtype} {norm.name} from a nested subproject.')
+
+        # Use os.path.abspath() to eliminate .. segments, but do not resolve symlinks
+        norm = Path(os.path.abspath(Path(srcdir, fname)))
+        do_validate_within_subproject(norm)
+        self.validated_cache.add(fname)
```

Comment:
How does a validation cache compare with `@functools.lru_cache(maxsize=None)`?

Reply:
I can change it but there are conceptual reasons why I didn't use lru_cache.

I didn't like, mostly, that lru_cache would create a tuple of all the arguments. The cache expresses that the function depends on subdir/fname, and ignores the interpreter because it's stored in its dictionary.

It's possible that lru_cache is a tiny bit *more* efficient than the handwritten cache actually, because a few checks in validate_within_subproject are done before the cache lookup in my code. Though more likely it's in the noise.

Reply:
I guess it cannot be an lru_cache on `do_validate_within_subproject` because we dynamically create the function... anyway your reasoning sounds ok, I think.